### ISSUE TITLE
Active file tab in editor gets moved to the right when tab bar is wider than editor pane

### ISSUE THREAD

**Issue by Khris777** [spyder-ide/spyder#5459]
## Description of your problem

Take the situation when you have lots of files open in the editor so that their tabs don't fit the width of the editor pane and you get those two arrows on the right side next to the tab bar with which you can scroll left or right through your open files.

If you now take any file that is not on the right side and edit it the whole tab bar scrolls horizontally to push the active tab to the right rim of the editor or as far as possible.

Is there a good reason for this? Personally I find this behaviour quite annoying.

**What is the expected output?**

I don't want to tab bar to scroll on its own only because I'm editing one of the files.

## Versions and main components

* Spyder Version: 3.2.3
* Python Version: 3.6.3
* Qt Version: 5.6.2
* PyQt Version: 5.6
* Operating system: Windows 10

## Dependencies

IPython >=4.0    :  6.2.1 (OK)
cython >=0.21    :  0.26.1 (OK)
jedi >=0.9.0     :  0.10.2 (OK)
nbconvert >=4.0  :  5.3.1 (OK)
numpy >=1.7      :  1.13.1 (OK)
pandas >=0.13.1  :  0.20.3 (OK)
pycodestyle >=2.3:  2.3.1 (OK)
pyflakes >=0.6.0 :  1.5.0 (OK)
pygments >=2.0   :  2.2.0 (OK)
pylint >=0.25    :  1.7.2 (OK)
qtconsole >=4.2.0:  4.3.1 (OK)
rope >=0.9.4     :  0.10.5 (OK)
sphinx >=0.6.6   :  1.6.3 (OK)
sympy >=0.7.3    :  1.1.1 (OK)

**Comment by ccordoba12** [comment]
This behavior was asked by several users in the past, so we changed how the Editor tab scrolls to fit it.

**Comment by jnsebgosselin** [comment]
@ccordoba12  I think this is a duplicate of Issue #1170, which should have been fixed by @rlaverde in PR #4607. However, I can confirm that I can reproduce the undesired behavior now on the master branch.

![tabs_auto_scrolling](https://user-images.githubusercontent.com/10170372/31548615-62c10d6e-aff9-11e7-8049-38c33fa98aae.gif)

**Comment by jnsebgosselin** [comment]
@ccordoba12 I was wrong. Issue #1170 and PR#4607 seems to relate to the undesired behaviour when **selecting** tabs, not when **editing** the corresponding file.

**Comment by ccordoba12** [comment]
@rlaverde, what do you think about this one?

**Comment by rlaverde** [comment]
I can confirm this issue too, is different to #1170 (but It's related)

This is caused because when writing a file,the test of the tab change (file--->file*),and that update cause the scrolling,but I don't know how It could be fixed (#4607 only fixes when selecting the tabs, avoiding the unnecessary update of the tabtext 725c0159f6371787e51998ca808e8c2cc723f5be)

**Comment by ccordoba12** [comment]
@rlaverde, any way to fix this one?

**Comment by jnsebgosselin** [comment]
@ccordoba12 I've looked a little bit into it and I think this is going to be hard to fix. This is occurring in private methods on the Qt side and I haven't found a way to prevent this undesired behavior. I would say this is a Qt bug per se... but if we wait for this to be fixed on the Qt side, we may wait for a long time.

I'll take a look again in case I've missed something on my first try. It's probably possible to go around it, but this is going to be some kind of a hack.

**Comment by rlaverde** [comment]
> any way to fix this one?

No, I don't know how it could be fixed, sorry 😞, as @jnsebgosselin says It looks like a Qt issue, I fixed the other one (#1170) with a workaround (not updating the tar title) but that workaround is not possible for this case 😕

**Comment by ccordoba12** [comment]
Ok, thanks for the info @jnsebgosselin and @rlaverde.

I'm closing this one for now. @jnsebgosselin, please reopen it if you find a workaround.

**Comment by bcolsen** [comment]
This was irritating me today. It is very difficult to work on two documents with the tabs moving around so I looked up some options. 

I seems to happen whenever the title of the tab changes. This is the worst when adding or removing  the "*" on changing or saving the document.

## Option 1
I thought we could see if changing the text color would force a redraw. Yellow for modified, red for read only

## Option 2 
Notepadqq changes icons in the tabs and the tabs don't move.

**Comment by bcolsen** [comment]
## Option 3
Let it move and move it back. Can we control the view at all?

**Comment by ccordoba12** [comment]
@jnsebgosselin, we could use your input here. What do you think about @bcolsen's ideas?

**Comment by jnsebgosselin** [comment]
I agree this is irritating for me too. I will take a look at it again with a fresh eye and see if I can find a solution.

**Comment by jnsebgosselin** [comment]
And if I can't find a solution, I think option 2 is the one I prefer.

**Comment by goanpeca** [comment]
It seems not much to do in that area

https://code.woboq.org/qt5/qtbase/src/widgets/widgets/qtabbar.cpp.html#_ZN14QTabBarPrivate7refreshEv
```cpp

void QTabBarPrivate::refresh()
--
{
Q_Q(QTabBar);
 
// be safe in case a subclass is also handling move with the tabs
if (pressedIndex != -1
&& movable
&& QGuiApplication::mouseButtons() == Qt::NoButton) {
moveTabFinished(pressedIndex);
if (!validIndex(pressedIndex))
pressedIndex = -1;
}
 
if (!q->isVisible()) {
layoutDirty = true;
} else {
layoutTabs();
makeVisible(currentIndex);
q->update();
q->updateGeometry();
}
}

```

**Comment by goanpeca** [comment]
>Let it move and move it back. Can we control the view at all?

I think we can;t... :-\

**Comment by jnsebgosselin** [comment]
For the sake of documenting this Issue, here are some relevant links:
https://bugreports.qt.io/browse/QTBUG-2360
https://bugreports.qt.io/browse/QTBUG-45381
https://stackoverflow.com/questions/32211741/qtabbar-to-preserve-tab-position
https://github.com/qutebrowser/qutebrowser/issues/3579

**Comment by bcolsen** [comment]
The more I think about it adding the "*"s actually the the work around for gui tool kits that don't allow icons on their tabs. I think that a simple small colored dot icon could convey as much meaning as a "*"

**Comment by jnsebgosselin** [comment]
> I thought we could see if changing the text color would force a redraw. Yellow for modified, red for read only

This would work and my vote goes to this option.

It also takes less horizontal space, so this is an additional benefit.

![tabs_moving](https://user-images.githubusercontent.com/10170372/69778503-8562a580-1172-11ea-9e1e-d75caa53106b.gif)

![tabs_moving_color](https://user-images.githubusercontent.com/10170372/69778514-8bf11d00-1172-11ea-9212-cb65861c4f77.gif)

**Comment by goanpeca** [comment]
This is awesome news @jnsebgosselin !

Could you post a version for the dark theme to see how it looks?

**Comment by jnsebgosselin** [comment]
Ah well... of course it doesn't work for qdarkstyle.

It seems like [QTabBar.setTabTextColor](https://doc.qt.io/qt-5/qtabbar.html#setTabTextColor) has no effect, probably because it is ignored by qdarkstyle during the painting.

**Comment by jnsebgosselin** [comment]
> 
> Ah well... of course it doesn't work for qdarkstyle.
> 
> It seems like [QTabBar.setTabTextColor](https://doc.qt.io/qt-5/qtabbar.html#setTabTextColor) has no effect, probably because it is ignored by qdarkstyle during the painting.

@goanpeca Do you think this would be relevant to open an issue on qdarkstyle repo?

**Comment by goanpeca** [comment]
>@goanpeca Do you think this would be relevant to open an issue on qdarkstyle repo?

I think we need to solve this on the Spyder side, but fear not, this is relatively easy, I know how to. Give me some minutes and I will add an example for you.

**Comment by jnsebgosselin** [comment]
Excellent! thanks

**Comment by goanpeca** [comment]
Ok, found an article that explains this @jnsebgosselin 

https://wiki.qt.io/Dynamic_Properties_and_Stylesheets

We could  define a property like `'dirty'` or `'unsaved'` and add this custom stylesheet. We could then  override the setTabTextColor to do the style()polish calls etc... if on darkstyle.

**Comment by jnsebgosselin** [comment]
> 
> Ok, found an article that explains this @jnsebgosselin
> 
> https://wiki.qt.io/Dynamic_Properties_and_Stylesheets
> 
> We could define a property like `'dirty'` or `'unsaved'` and add this custom stylesheet and append to the normal stylesheet. We could then override the setTabTextColor to do the style()polish calls etc... if on darkstyle.

Thank you, but I don't understand how we can apply this to each tab of the QTabBar individually. I don't see a way to access the tabs from the public API.

**Comment by goanpeca** [comment]
You are absolutely right, tabs are not widgets... internally structs that hold info so this will not work :-\ 

We should probably remove any Style that QdarkStyle applies to text color and see if that way we can still control the color with the method. Otherwise we would need to reimplement the paint method.

https://code.woboq.org/qt5/qtbase/src/widgets/widgets/qtabbar.cpp.html#_ZN7QTabBar10paintEventEP11QPaintEvent

or something like https://gist.github.com/eyllanesc/4a821d2956b5e2ee841f3eb156da078d

**Comment by jnsebgosselin** [comment]
> We should probably remove any Style that QdarkStyle applies to text color and see if that way we can still control the color with the method.

I have try that already but failed. I will try harder and see if I can make it work lol

> Otherwise we would need to reimplement the paint method.

That would be though though, especially since we support dragging and moving... I say we should try to avoid doing this if we can help it.

My favorite option would be to fix this upstream in qdarkstyle, but I don't know if this is possible.

**Comment by goanpeca** [comment]
>I have try that already but failed. I will try harder and see if I can make it work lol

I can have a look at it too :-)

**Comment by jnsebgosselin** [comment]
> 
> 
> > I have try that already but failed. I will try harder and see if I can make it work lol
> 
> I can have a look at it too :-)

Oh yes, please do, that would be greatly appreciated, thanks.

**Comment by goanpeca** [comment]
Ok, yes we need to disable a change on QDark and this should work:

```python
qss = '''
QTabBar::tab {
    color: green;
}
```
As soon as we set `color` css property, we override the method.
'''

**Comment by goanpeca** [comment]
@jnsebgosselin please open an issue so I can track it! thanks :-)

**Comment by jnsebgosselin** [comment]
> 
> 
> @jnsebgosselin please open an issue so I can track it! thanks :-)
> 
> https://github.com/ColinDuquesnoy/QDarkStyleSheet

Excellent, thank you! I will open an issue on qdark about this then with a MWE.

**Comment by jnsebgosselin** [comment]
The relevant issue in qdarstyle is here: https://github.com/ColinDuquesnoy/QDarkStyleSheet/issues/212

**Comment by ccordoba12** [comment]
This hasn't been fixed, sorry.

**Comment by jnsebgosselin** [comment]
@isabela-pf  what do you think about this issue? This is indeed quite irritating I must admit. However, there is technical limitations in Qt that prevent us from fixing this "bug" while keeping the current behaviour intact.

Therefore, we were thinking about changing the text color of the tabs instead of adding a `*` at the end of the name to indicate unsaved changes. However a technical limitation in `qdarstyle` at the time (which has now been fixed) prevented us from implementing this solution.

What do you think? You can see some animations of what I am talking about here https://github.com/spyder-ide/spyder/issues/5459#issuecomment-559343400.

**Comment by isabela-pf** [comment]
Hello there! I think with the limitations, the icon approach does make the most sense. In general, I wouldn't recommend using only a change in color to indicate state so I don't think changing the text color will work well. Even though it takes up more space in the tag, I think the approach that has icons showing the state makes the most sense. 

If it doesn't force the redraw, the best case would be to switch icons having the check for saved and the circle for unsaved changes. We can have colors too if wanted. Something kind of like this:
<img width="704" alt="0-2" src="https://user-images.githubusercontent.com/50221806/115782845-01b1e600-a371-11eb-8296-d4d1c67e87b1.png">
<img width="704" alt="0-3" src="https://user-images.githubusercontent.com/50221806/115782848-024a7c80-a371-11eb-93d5-89d4d2255979.png">

Does that make sense or seem feasible?

**Comment by jnsebgosselin** [comment]
Thank you @isabela-pf.

Yes that makes sense and it is also feasible (see animation below). It is also a change that is not really complicated to implement in the code. Please don't mind the icons and colors that I used in the animation, it was only to show that is is working.

![tab_saved_state](https://user-images.githubusercontent.com/10170372/115786377-bc041600-a38e-11eb-827a-a732bd409b66.gif)

**Comment by jnsebgosselin** [comment]
The only downside is that we cannot adjust the space between the icon and the text.

We could, but not in a pretty and maintainable way I think. We would have to subclass QTabBar and paint the tabs ourselves, which is something we should try to avoid as much as possible I think.

![image](https://user-images.githubusercontent.com/10170372/115882940-41cda300-a41b-11eb-8263-920586217b64.png)

**Comment by ccordoba12** [comment]
Yeah, that's really not good.

**Comment by jnsebgosselin** [comment]
> Yeah, that's really not good.

I agree it makes the tabs quite larger.

I think I have an idea how we can solve this and preserve the current behaviour. I will report back when I have the time to test it.

**Comment by jnsebgosselin** [comment]
Ok well, this doesn't work. There is too much stuff going on in private functions on the Qt side that are not exposed on the Python side for us to solve this easily. So I I guess we have the following options:

1 ) Change the color of the text only to indicate unsaved changes.
2) Add an icon and accept that we won't be able to control where the icon is placed in the tab.
3) Subclass QTabBar and paint the tabs AND the moving tab ourselves so that we can do whatever we want to fix this issue.

I can work on 3 if option 1 and 2 are not acceptable to solve this issue. I am not sure it would be too hard to do. All the logic is already there. We just have to reimplement it in Python.

I am personally using option 2 locally and though it is true that it makes the tabs substantially larger, it is still way better than the current behaviour.

**Comment by ccordoba12** [comment]
> 3) Subclass QTabBar and paint the tabs AND the moving tab ourselves so that we can do whatever we want to fix this issue.

I'd say we should go with this option because it seems to be the most sustainable for the future.

**Comment by jnsebgosselin** [comment]
Ok this what I would have picked too. I'll see what I can do.

Meanwhile, maybe you can continue to work out the UI/UX details with @isabela-pf.

**Comment by ccordoba12** [comment]
> Meanwhile, maybe you can continue to work out the UI/UX details with @isabela-pf.

Oh, sorry for not mentioning anything about that. I agree with using different colored icons, as @isabela-pf suggested in https://github.com/spyder-ide/spyder/issues/5459#issuecomment-825173564.

So I guess it would be a matter of adding the changes you already have to the new QTabBar subclass you plan to implement.

**Comment by jeremymatt** [comment]
Well, it seems that I'm not the only person who finds this behavior irritating, which is comforting.  Thanks for everyone's work on this.  FWIW, I'd be happy with any solution that fixes this, even if it substantially increases tab sizes.  I'm generally just flipping back and forth between two or three files so larger tabs wouldn't be an irritation for me.

As a workaround, taking a moment to sort open tabs by activity such that the most active tabs are at the left of the tabbar reduces the number of times this is encountered.

**Comment by jnsebgosselin** [comment]
I am going to try to submit a PR for this issue soon. I should have the time to work on this in the following weeks.

**Comment by jnsebgosselin** [comment]
Ok, I've looked into this issue (again) and unless I am missing something, we basically would have to rewrite almost the whole `QTabBar` class in Python in order to paint the tabs ourselves. `paintEvent` depends on a lot of private methods, which in turn depend on a lot of private attributes that are defined all over the place in `QTabBar`.

What do you think @ccordoba12? It seems like going that route would be a lot of work and would add a lot of maintenance overhead in order to fix a, relatively speaking, minor UX inconvenience that should, imho, truly be fixed on the Qt side instead.

Maybe this is already or is going to be fixed in Qt6, there seems to be a lot of refactoring going on lately with the QTabBar class.
https://github.com/qt/qtbase/commits/5d4b91ea66666f3c8d3c8640b87a33c13dbd0c2c/src/widgets/widgets/qtabbar.cpp

Even if its not fixed in Qt6, in PyQt6, it seems like `initStyleOption` has been made virtual. So I think we might be able to easily implement a workaround by extending `QTabBar.initStyleOption` in order to add an `*` to the tabs that are modified instead of calling `QTabBar.setTabText` directly.

For Qt5 and PyQt5 though, I think our best option is to either do nothing or add an option to add an icon like I did in https://github.com/spyder-ide/spyder/issues/5459#issuecomment-825193971, even though this makes the tabs wider.

**Comment by jnsebgosselin** [comment]
This bug is more a Qt than a Spyder bug per se and we have not been able to reach consensus to work around this issue in Spyder.

However, I think this annoying behavior was fixed in Qt 6.5.

See :
https://bugreports.qt.io/browse/QTBUG-45381
https://github.com/qt/qtbase/commit/e851d4c06154bf02b23030ff1f7024a8b9edf874

**Comment by jnsebgosselin** [comment]
The option to add an icon makes the tabs substantially larger and was not accepted because of that. The option to simply change the text color was not accepted for accessibility reason I think.

**Comment by jnsebgosselin** [comment]
I know this is not ideal, but you can easily patch this in your Spyder version by changing just a few lines of codes. Something like this:

![image](https://github.com/spyder-ide/spyder/assets/10170372/a9107a22-6504-4d27-b140-8af643b58d7f)

**Comment by ccordoba12** [comment]
In PR #21276 I added our own close button for tabs in the editor and IPython console. With that we'll be able to implement the same UI as VSCode, i.e. showing a filled circle instead of an `x` for unsaved files. I think that will avoid the current behavior until we're able to support Qt 6.

**Comment by ptth222** [comment]
> I know this is not ideal, but you can easily patch this in your Spyder version by changing just a few lines of codes. Something like this:
> 
> ![image](https://private-user-images.githubusercontent.com/10170372/289934595-a9107a22-6504-4d27-b140-8af643b58d7f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk4NzMzNzMsIm5iZiI6MTcxOTg3MzA3MywicGF0aCI6Ii8xMDE3MDM3Mi8yODk5MzQ1OTUtYTkxMDdhMjItNjUwNC00ZDI3LWIxNDAtOGFmNjQzYjU4ZDdmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAxVDIyMzExM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkMmYxMjFiMWU4YzAwZTBjOGQwNTUwMzcyMTc0NTFmOGRiOGRiNDg4Zjk1MmNlZDliYjVkZWFlNDFiM2UzOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5d6LUysxsTwcvEHBhIqycD9fyZn5YWD3iFtmImYPeok)

This worked for me. Red and black were not good choices for the dark theme IMO.  I went with something closer to orange. Here is the code I used:

```
from qtpy.QtGui import QColor
self.tabs.tabBar().setTabTextColor(index, QColor('#f5a742') if is_modified else QColor('white'))
```

It sounds like Qt 6 might finally fix this issue. If not, another good workaround I think would be to add an option on right-click or in the "Options" on the right side of the tab bar to send the tab left. The reposition doesn't happen if the tab bar is all the way on the left. I will often manually move all of my most active tabs to the left to avoid the repositioning. Being able to right-click and "Move Left" would make this less painful. Being able to do it on on the inactive tabs would be great as well, so you wouldn't have to "Move Left" and then scroll back right to move the next one left. Of course, now I think this color thing is going to be fine for me.
